File: static/js/ledger.js
// ledger.js - 総勘定元帳

let accountsList = [];
let selectedAccountId = null;
let editingTid = null;

const ELEMENT_LABELS = {
  assets:      '資産',
  liabilities: '負債',
  equity:      '純資産',
  revenues:    '収益',
  expenses:    '費用',
};
const ELEMENT_ORDER = ['assets', 'liabilities', 'equity', 'revenues', 'expenses'];

function fmt(n) {
  return Number(n).toLocaleString('ja-JP');
}

function escHtml(s) {
  return String(s)
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;');
}

function showAlert(msg, type) {
  $('#alert-box').html(
    `<div class="alert alert-${type} alert-dismissible"><button type="button" class="btn-close" data-bs-dismiss="alert"></button>${msg}</div>`
  );
}

// ---- 科目ツリー（サイドバー） ----

function loadAccounts() {
  return $.getJSON('/api/accounts').then(function(data) {
    accountsList = data;
    renderSidebar(data);
    renderMobileSelect(data);
  });
}

function renderSidebar(accounts) {
  const grouped = {};
  accounts.forEach(function(a) {
    if (!grouped[a.element]) grouped[a.element] = [];
    grouped[a.element].push(a);
  });

  let html = '';
  ELEMENT_ORDER.forEach(function(el) {
    if (!grouped[el] || grouped[el].length === 0) return;
    html += `<div class="text-muted small px-2 pt-2 pb-1 fw-bold">${escHtml(ELEMENT_LABELS[el] || el)}</div>`;
    grouped[el].forEach(function(a) {
      html += `<div class="px-2 py-1 rounded account-item ${a.id === selectedAccountId ? 'bg-primary text-white' : ''}"
                   style="cursor:pointer;font-size:.9rem" data-id="${a.id}">${escHtml(a.name)}</div>`;
    });
  });
  $('#account-tree').html(html || '<div class="text-muted small p-2">科目がありません</div>');
}

function renderMobileSelect(accounts) {
  const sel = $('#account-select-mobile');
  sel.empty().append('<option value="">科目を選択してください</option>');
  const grouped = {};
  accounts.forEach(function(a) {
    if (!grouped[a.element]) grouped[a.element] = [];
    grouped[a.element].push(a);
  });
  ELEMENT_ORDER.forEach(function(el) {
    if (!grouped[el] || grouped[el].length === 0) return;
    const og = $(`<optgroup label="${escHtml(ELEMENT_LABELS[el] || el)}">`);
    grouped[el].forEach(function(a) {
      og.append(`<option value="${a.id}">${escHtml(a.name)}</option>`);
    });
    sel.append(og);
  });
  if (selectedAccountId) sel.val(selectedAccountId);
}

function selectAccount(accountId) {
  selectedAccountId = accountId;

  // サイドバーのハイライト更新
  $('#account-tree .account-item').each(function() {
    const isSelected = parseInt($(this).data('id'), 10) === accountId;
    $(this).toggleClass('bg-primary text-white', isSelected);
  });

  // モバイルセレクト同期
  $('#account-select-mobile').val(accountId);

  const acc = accountsList.find(a => a.id === accountId);
  $('#selected-account-name').text(acc ? acc.name : '');
  $('#btn-export').prop('disabled', false);

  loadLedger();
}

// ---- 元帳データ取得・表示 ----

function loadLedger() {
  if (!selectedAccountId) return;
  const from = $('#filter-from').val();
  const to   = $('#filter-to').val();

  const params = { account_id: selectedAccountId };
  if (from) params.from = from;
  if (to)   params.to   = to;

  $('#ledger-area').html('<div class="text-muted p-3">読み込み中...</div>');

  $.getJSON('/api/ledger', params).then(function(data) {
    renderLedger(data);
  }).fail(function(xhr) {
    $('#ledger-area').html(
      `<div class="text-danger p-3">${escHtml(xhr.responseJSON?.error || '読み込みに失敗しました')}</div>`
    );
  });
}

function renderLedger(data) {
  const entries = data.entries;
  const ob      = data.opening_balance;

  let html = '<div class="table-responsive"><table class="table table-sm table-bordered table-hover mb-0">';
  html += `<thead class="table-light">
    <tr>
      <th style="width:100px">日付</th>
      <th>備考</th>
      <th class="text-end" style="width:110px">借方</th>
      <th class="text-end" style="width:110px">貸方</th>
      <th class="text-end" style="width:120px">残高</th>
      <th class="d-none d-md-table-cell" style="width:120px">相手科目</th>
    </tr>
  </thead><tbody>`;

  // 前期残高行
  if (data.from_date && ob !== 0) {
    html += `<tr class="table-secondary">
      <td>${escHtml(data.from_date)}</td>
      <td class="text-muted fst-italic">（前期残高）</td>
      <td></td><td></td>
      <td class="text-end">${fmt(ob)}</td>
      <td class="d-none d-md-table-cell"></td>
    </tr>`;
  }

  if (entries.length === 0 && ob === 0) {
    html += `<tr><td colspan="6" class="text-muted text-center p-3">仕訳がありません</td></tr>`;
  }

  entries.forEach(function(e) {
    const cps = e.counterparts;
    const cpStr = cps.length === 0 ? '' : (cps.length === 1 ? cps[0] : '諸口');
    html += `<tr style="cursor:pointer" class="ledger-row" data-tid="${escHtml(e.transaction_id)}">
      <td>${escHtml(e.entry_date)}</td>
      <td>${escHtml(e.note)}</td>
      <td class="text-end">${e.debit  ? fmt(e.debit)  : ''}</td>
      <td class="text-end">${e.credit ? fmt(e.credit) : ''}</td>
      <td class="text-end">${fmt(e.balance)}</td>
      <td class="d-none d-md-table-cell text-muted small">${escHtml(cpStr)}</td>
    </tr>`;
  });

  html += '</tbody></table></div>';
  $('#ledger-area').html(html);
}

// ---- TSV エクスポート ----

$('#btn-export').on('click', function() {
  if (!selectedAccountId) return;
  const from = $('#filter-from').val();
  const to   = $('#filter-to').val();
  let url = `/api/export/ledger?account_id=${selectedAccountId}`;
  if (from) url += `&from=${encodeURIComponent(from)}`;
  if (to)   url += `&to=${encodeURIComponent(to)}`;
  window.location.href = url;
});

// ---- 表示ボタン ----

$('#btn-filter').on('click', function() {
  loadLedger();
});

// ---- イベント：サイドバー科目クリック ----

$(document).on('click', '.account-item', function() {
  selectAccount(parseInt($(this).data('id'), 10));
});

// ---- イベント：モバイルセレクト ----

$('#account-select-mobile').on('change', function() {
  const id = parseInt($(this).val(), 10);
  if (id) selectAccount(id);
});

// ---- イベント：元帳行クリック → 編集モーダル ----

$(document).on('click', '.ledger-row', function() {
  const tid = $(this).data('tid');
  openEditModal(tid);
});

// ---- 編集モーダル ----

function accountOptions(selectedId) {
  return accountsList.map(function(a) {
    const sel = a.id === selectedId ? 'selected' : '';
    return `<option value="${a.id}" ${sel}>${escHtml(a.name)}</option>`;
  }).join('');
}

function addEditRow(dc, accountId, amount) {
  const row = $(`
    <tr>
      <td>
        <select class="form-select form-select-sm edit-dc">
          <option value="debit"  ${dc === 'debit'  ? 'selected' : ''}>借方</option>
          <option value="credit" ${dc === 'credit' ? 'selected' : ''}>貸方</option>
        </select>
      </td>
      <td>
        <select class="form-select form-select-sm edit-account">${accountOptions(accountId)}</select>
      </td>
      <td>
        <input type="number" class="form-control form-control-sm edit-amount" min="1" step="1" value="${amount || ''}">
      </td>
      <td>
        <button type="button" class="btn btn-sm btn-outline-danger edit-remove-row">×</button>
      </td>
    </tr>
  `);
  $('#edit-lines-body').append(row);
}

function updateEditBalance() {
  let debit = 0, credit = 0;
  $('#edit-lines-body tr').each(function() {
    const dc  = $(this).find('.edit-dc').val();
    const amt = parseInt($(this).find('.edit-amount').val(), 10);
    if (!isNaN(amt) && amt > 0) {
      if (dc === 'debit') debit += amt;
      else credit += amt;
    }
  });
  if (debit === 0 && credit === 0) {
    $('#edit-balance').html('');
    return;
  }
  const ok  = debit > 0 && debit === credit;
  const cls = ok ? 'text-success' : 'text-danger';
  $('#edit-balance').html(
    `借方合計 / 貸方合計: <span class="${cls} fw-bold">${fmt(debit)} / ${fmt(credit)}</span>`
  );
}

function openEditModal(tid) {
  $.getJSON(`/api/journal/transaction/${tid}`).then(function(tx) {
    editingTid = tx.transaction_id;
    $('#edit-date').val(tx.entry_date);
    $('#edit-note').val(tx.note || '');
    $('#edit-lines-body').empty();
    tx.lines.forEach(function(line) {
      addEditRow(line.debit_credit, line.account_id, line.amount);
    });
    updateEditBalance();
    new bootstrap.Modal('#editTxnModal').show();
  }).fail(function() {
    showAlert('取引の読み込みに失敗しました', 'danger');
  });
}

$(document).on('change input', '#edit-lines-body .edit-dc, #edit-lines-body .edit-amount', updateEditBalance);

$(document).on('click', '.edit-remove-row', function() {
  if ($('#edit-lines-body tr').length <= 2) {
    alert('最低2行必要です');
    return;
  }
  $(this).closest('tr').remove();
  updateEditBalance();
});

$('#edit-add-row').on('click', function() {
  addEditRow('debit', accountsList[0]?.id, '');
  updateEditBalance();
});

$('#btn-edit-save').on('click', function() {
  const entry_date = $('#edit-date').val();
  const note       = $('#edit-note').val().trim();
  if (!entry_date) { alert('取引日を入力してください'); return; }

  const lines = [];
  let valid = true;
  $('#edit-lines-body tr').each(function() {
    const dc         = $(this).find('.edit-dc').val();
    const account_id = parseInt($(this).find('.edit-account').val(), 10);
    const amount     = parseInt($(this).find('.edit-amount').val(), 10);
    if (isNaN(amount) || amount <= 0) { alert('金額を正しく入力してください'); valid = false; return false; }
    lines.push({ account_id, debit_credit: dc, amount });
  });
  if (!valid) return;

  $.ajax({
    url: `/api/journal/transaction/${editingTid}`,
    method: 'PUT',
    contentType: 'application/json',
    data: JSON.stringify({ entry_date, note, lines }),
    success: function() {
      bootstrap.Modal.getInstance('#editTxnModal').hide();
      showAlert('更新しました', 'success');
      loadLedger();
    },
    error: function(xhr) {
      alert(xhr.responseJSON?.error || '更新に失敗しました');
    }
  });
});

$('#btn-edit-delete').on('click', function() {
  if (!confirm('この取引を削除しますか？\n（この操作は取り消せません）')) return;
  $.ajax({
    url: `/api/journal/transaction/${editingTid}`,
    method: 'DELETE',
    success: function() {
      bootstrap.Modal.getInstance('#editTxnModal').hide();
      showAlert('削除しました', 'success');
      loadLedger();
    },
    error: function(xhr) {
      showAlert(xhr.responseJSON?.error || '削除に失敗しました', 'danger');
    }
  });
});

// ---- 初期化 ----

$(function() {
  loadAccounts();
});

File: blueprints/ledger.py
# -*- coding: utf-8 -*-
from collections import defaultdict
from flask import Blueprint, request, jsonify, render_template
from flask_login import login_required
from helpers import get_db, db_required, tsv_response

ledger_bp = Blueprint('ledger', __name__)

_DEBIT_NORMAL = ('assets', 'expenses')


@ledger_bp.get('/ledger')
@login_required
@db_required
def ledger():
    return render_template('ledger.html')


def _build_ledger(db, account_id, from_date, to_date):
    """
    元帳データを構築して返す。

    戻り値: (data_dict, None) または (None, 'not_found')
    """
    acc = db.execute(
        'SELECT id, name, element FROM accounts WHERE id=?', (account_id,)
    ).fetchone()
    if not acc:
        return None, 'not_found'

    is_debit_normal = acc['element'] in _DEBIT_NORMAL

    # --- 前期残高（from_date より前の累計） ---
    if from_date:
        ob = db.execute('''
            SELECT COALESCE(SUM(
                CASE WHEN debit_credit='debit' THEN amount ELSE -amount END
            ), 0) as net
            FROM journal WHERE account_id=? AND entry_date < ?
        ''', (account_id, from_date)).fetchone()
        raw = ob['net'] or 0
        opening_balance = raw if is_debit_normal else -raw
    else:
        opening_balance = 0

    # --- 対象期間の仕訳行 ---
    sql = '''
        SELECT j.id, j.entry_date, j.transaction_id,
               j.debit_credit, j.amount, j.note
        FROM journal j
        WHERE j.account_id = ?
    '''
    params = [account_id]
    if from_date:
        sql += ' AND j.entry_date >= ?'
        params.append(from_date)
    if to_date:
        sql += ' AND j.entry_date <= ?'
        params.append(to_date)
    sql += ' ORDER BY j.entry_date, j.id'

    rows = db.execute(sql, params).fetchall()

    # --- 相手科目を一括取得（N+1 回避） ---
    tids = list(dict.fromkeys(r['transaction_id'] for r in rows))
    counterpart_map = defaultdict(list)
    if tids:
        ph = ','.join('?' * len(tids))
        cp_rows = db.execute(f'''
            SELECT j.transaction_id, a.name
            FROM journal j JOIN accounts a ON j.account_id = a.id
            WHERE j.transaction_id IN ({ph}) AND j.account_id != ?
            ORDER BY j.id
        ''', tids + [account_id]).fetchall()
        for r in cp_rows:
            counterpart_map[r['transaction_id']].append(r['name'])

    # --- 残高を累積しながらエントリ構築 ---
    balance = opening_balance
    entries = []
    for row in rows:
        raw_delta = row['amount'] if row['debit_credit'] == 'debit' else -row['amount']
        delta = raw_delta if is_debit_normal else -raw_delta
        balance += delta

        # 相手科目（重複除去・順序保持）
        counterparts = list(dict.fromkeys(counterpart_map.get(row['transaction_id'], [])))

        entries.append({
            'entry_date':     row['entry_date'],
            'transaction_id': row['transaction_id'],
            'note':           row['note'] or '',
            'debit':          row['amount'] if row['debit_credit'] == 'debit'  else 0,
            'credit':         row['amount'] if row['debit_credit'] == 'credit' else 0,
            'balance':        balance,
            'counterparts':   counterparts,
        })

    return {
        'account':         {'id': acc['id'], 'name': acc['name'], 'element': acc['element']},
        'from_date':       from_date,
        'to_date':         to_date,
        'opening_balance': opening_balance,
        'entries':         entries,
    }, None


@ledger_bp.get('/api/ledger')
@login_required
@db_required
def api_ledger():
    account_id = request.args.get('account_id', type=int)
    from_date  = request.args.get('from', '').strip()
    to_date    = request.args.get('to',   '').strip()

    if not account_id:
        return jsonify({'error': '科目を指定してください'}), 400

    data, err = _build_ledger(get_db(), account_id, from_date, to_date)
    if err:
        return jsonify({'error': '科目が見つかりません'}), 404
    return jsonify(data)


@ledger_bp.get('/api/export/ledger')
@login_required
@db_required
def api_export_ledger():
    account_id = request.args.get('account_id', type=int)
    from_date  = request.args.get('from', '').strip()
    to_date    = request.args.get('to',   '').strip()

    if not account_id:
        return jsonify({'error': '科目を指定してください'}), 400

    db = get_db()
    acc = db.execute('SELECT code FROM accounts WHERE id=?', (account_id,)).fetchone()
    if not acc:
        return jsonify({'error': '科目が見つかりません'}), 404

    data, err = _build_ledger(db, account_id, from_date, to_date)
    if err:
        return jsonify({'error': '科目が見つかりません'}), 404

    headers = ['日付', '備考', '取引番号', '借方', '貸方', '残高', '相手科目']
    rows = []

    if from_date and data['opening_balance'] != 0:
        rows.append([from_date, '（前期残高）', '', '', '', data['opening_balance'], ''])

    for e in data['entries']:
        cps = e['counterparts']
        cp_str = cps[0] if len(cps) == 1 else ('諸口' if cps else '')
        rows.append([
            e['entry_date'],
            e['note'],
            e['transaction_id'],
            e['debit']  or '',
            e['credit'] or '',
            e['balance'],
            cp_str,
        ])

    code = acc['code']
    if from_date or to_date:
        f = from_date or 'start'
        t = to_date   or 'end'
        filename = f'ledger_{code}_{f}_{t}.tsv'
    else:
        filename = f'ledger_{code}_all.tsv'

    return tsv_response(rows, headers, filename)
